File: apps/csharp/tests/Ch08/BankAccountTests.cs
using GrokkingConcurrency.CSharp.Ch08;

namespace GrokkingConcurrency.CSharp.Tests.Ch08;

/// <summary>
/// 銀行口座のテスト（同期と排他制御）
/// </summary>
public class BankAccountTests
{
    [Fact]
    public void ShouldDeposit()
    {
        var account = new BankAccount(100);

        account.Deposit(50);

        Assert.Equal(150, account.Balance);
    }

    [Fact]
    public void ShouldWithdraw()
    {
        var account = new BankAccount(100);

        var result = account.Withdraw(50);

        Assert.True(result);
        Assert.Equal(50, account.Balance);
    }

    [Fact]
    public void ShouldNotWithdrawWithInsufficientFunds()
    {
        var account = new BankAccount(100);

        var result = account.Withdraw(150);

        Assert.False(result);
        Assert.Equal(100, account.Balance);
    }

    [Fact]
    public void ShouldMaintainBalanceUnderConcurrentAccess()
    {
        var account = new BankAccount(1000);
        const int numThreads = 100;
        const int depositsPerThread = 10;
        using var countdown = new CountdownEvent(numThreads);

        var tasks = Enumerable.Range(0, numThreads)
            .Select(_ => Task.Run(() =>
            {
                for (var j = 0; j < depositsPerThread; j++)
                {
                    account.Deposit(1);
                }
                countdown.Signal();
            }))
            .ToArray();

        countdown.Wait(TimeSpan.FromSeconds(10));
        Task.WaitAll(tasks);

        // 1000 + (100 * 10) = 2000
        Assert.Equal(2000, account.Balance);
    }

    [Fact]
    public void ShouldTransferBetweenAccounts()
    {
        var from = new BankAccount(100);
        var to = new BankAccount(50);

        var result = BankAccount.Transfer(from, to, 30);

        Assert.True(result);
        Assert.Equal(70, from.Balance);
        Assert.Equal(80, to.Balance);
    }
}

File: apps/fsharp/tests/BankAccountTests.fs
module BankAccountTests

open Xunit
open System.Threading
open GrokkingConcurrency.Ch08.BankAccount

[<Fact>]
let ``BankAccount deposits correctly`` () =
    let account = createAccount 100
    deposit account 50
    Assert.Equal(150, getBalance account)

[<Fact>]
let ``BankAccount withdraws correctly`` () =
    let account = createAccount 100
    let result = withdraw account 50
    Assert.True(result)
    Assert.Equal(50, getBalance account)

[<Fact>]
let ``BankAccount does not withdraw more than balance`` () =
    let account = createAccount 100
    let result = withdraw account 150
    Assert.False(result)
    Assert.Equal(100, getBalance account)

[<Fact>]
let ``BankAccount transfer is atomic`` () =
    let from = createAccount 100
    let toAccount = createAccount 0

    let result = transfer from toAccount 50
    Assert.True(result)
    Assert.Equal(50, getBalance from)
    Assert.Equal(50, getBalance toAccount)

[<Fact>]
let ``BankAccount handles concurrent transfers safely`` () =
    let account1 = createAccount 1000
    let account2 = createAccount 1000

    let latch = new CountdownEvent(100)

    let tasks =
        [ for _ in 1 .. 50 do
            yield async {
                transfer account1 account2 10 |> ignore
                latch.Signal() |> ignore
            }
            yield async {
                transfer account2 account1 10 |> ignore
                latch.Signal() |> ignore
            } ]

    tasks
    |> Async.Parallel
    |> Async.RunSynchronously
    |> ignore

    latch.Wait()

    // Total balance should be preserved
    Assert.Equal(2000, getBalance account1 + getBalance account2)
